libzfs.pyx

```diff
@@ -2991,6 +3083,169 @@ cdef class ZFSDataset(ZFSResource):
+                        and child.key_location != 'prompt'
+                    ):
+                        dataset = child
+                        with nogil:
+                            ret = libzfs.zfs_crypto_load_key(dataset.handle, noop, alt_keylocation)
+                        if ret != 0:
+                            failed.append(self.root.get_error())
+                        tried += 1
+            finally:
+                if temp_file and os.path.exists(temp_file.name):
+                    os.unlink(temp_file.name)
+
+            self.root.write_history(
+                'zfs load-key', '-r' if recursive else '', f'-L {key_location}' if key_location else '',
+                '-n' if noop else '', self.name
+            )
+
+            if failed:
+                message = '\n'.join(f'{e.code}{f": {e.args[0]}" if e.args else ""}' for e in failed)
+                if recursive:
+                    message += f'\n{tried - len(failed)}/{tried} key(s) successfully loaded'
+                raise ZFSException(Error.CRYPTO_FAILED, message)
+
+        def load_key(self, recursive=False, key=None, key_location=None):
+            self.load_key_common(recursive, key_location, key, no_op=False)
+
+        def check_key(self, key=None, key_location=None):
+            try:
+                self.load_key_common(False, key_location, key, no_op=True)
+            except ZFSException:
+                return False
+            else:
+                return True
+
+        def unload_key(self, recursive=False):
+            cdef ZFSDataset dataset
+            failed = []
+            tried = 0
+            for child in itertools.chain([self], self.children_recursive if recursive else []):
+                if (child.encryption_root == child and child.key_loaded) or (child == self and not recursive):
+                    dataset = child
+                    with nogil:
+                        ret = libzfs.zfs_crypto_unload_key(dataset.handle)
+                    if ret != 0:
+                        failed.append(self.root.get_error())
+                    tried += 1
+
+            self.root.write_history('zfs unload-key', '-r' if recursive else '', self.name)
+
+            if failed:
+                message = '\n'.join(f'{e.code}{f": {e.args[0]}" if e.args else ""}' for e in failed)
+                if recursive:
+                    message += f'\n{tried - len(failed)}/{tried} key(s) successfully unloaded'
+                raise ZFSException(Error.CRYPTO_FAILED, message)
+
+        def change_key(self, props=None, load_key=False, inherit=False, key=None):
+            if not self.encrypted:
+                raise ZFSException(py_errno.EINVAL, f'{self.name} is not encrypted')
+
+            props = props or {}
+            if props and inherit:
+                raise ZFSException(py_errno.EINVAL, 'Properties not allowed for inheriting')
+            elif inherit:
+                if self.encryption_root != self:
```

Comment:
`self.encryption_root != self` is this correct? Looks weird at first glance

Reply:
Yup, `encryption_root` returns a `ZFSDataset` object or `None` if there is none